web/book/src/internals/type-system.md

```diff
@@ -308,25 +311,105 @@ Language and does not have to bother with exact physical layout of types.
 ### Non-bijection cases between PTS and STS
 
-There are cases where a PTS construct has multiple possible and valid
-representations in some STSs.
+### Type representations
+
+There are cases where a PTS type has multiple possible and valid representations
+in some STSs.
+
+For such cases, we'd want to support the use of alternative representations for
+storing data, but also application of any function that is defined for the
+original type.
+
+Using SQLite as an example again, users may have some temporal data stored as
+INTEGER unix timestamp and some as TEXT that contains ISO 8601 without timezone.
+From the user's perspective, both of these types are `timestamp`s and should be
+declared as such. But when compiling operations over these types to SQL, the
+compiler should consider their different representations in STS. For example a
+difference between two timestamps `timestamp - timestamp` can be translated to a
+normal int subtraction for INTEGER repr, but must apply SQLite's function
+`unixepoch` when dealing with TEXT repr.
+
+Table declarations should therefore support annotations that give hints about
+which representation is used:
+
+```
+table foo {
+    #[repr(text)]
+    created_at: timestamp,
+}
+```
+
+A similar example is an "array of strings type" in PTS that could be represented
+by a `text[]` (if DBMS supports arrays) or `json` or it's variant `jsonb` in
+Postgres. Again, the representation would affect operators: in Postgres, arrays
+would be accessed with `my_array[1]` and json arrays would use
+`my_json_array -> 1`. This example may not be applicable, if we decide that we
+want a separate JSON type in PST.
+
+### RQ functions, targets and reprs
+
+> This part is talks about technical implementations, not the language itself
+
+#### Idea
+
+RQ contains a single node kind for expressing operations and functions:
+BuiltInFunction (may be renamed in the future).
+
+It is a bottleneck that we can leverage when trying to affect how an operator or
+a function interacts with different type representations on different targets.
+
+Idea is to implement the BuiltInFunction multiple times and annotate it with it
+intended target and parameter representation. Then we can teach the compiler to
+pick the appropriate function implementation that suit current repr and
+compilation target.
+
+#### Specifics
+
+RQ specification is an interface that contains functions, identified by name
+(i.e. `std.int8.add`). These functions have typed parameters and a return value.
+If an RQ function call does not match the function declaration in number or in
+types of the parameters, this is considered an invalid RQ AST.
+
+We provide multiple implementations for each RQ function. They are annotated
+with a target (i.e. `#[target(sql.sqlite)]`) and have their params annotated
+with type reprs (i.e. `#[repr(int)]`).
+
+```
+# using a made-up syntax
+
+#[target(sql.sqlite)]
+func std.int8.add
+    #[repr(int8)] x
+    #[repr(int8)] y
+    -> s"{x} + {y}"
```

Comment:
Is there an example of when this would be useful? Maybe `in` for ranges vs. lists?

The ints example below is maybe OK, but not that tethered to what's likely to be helpful for PRQL. No objection to including for the moment, but future versions might want to more closely connect the abstract ideas to something concrete

Reply:
SQLite's temporal types.

Arrays (which have many different flavors in different dialects)